### DISCUSSION TITLE
Normals, tangents, and bitangents

### DISCUSSION THREAD

**Discussion by jwwalker** [google/filament#3463] [Q&A] [answered]
I see that one of the types of `VertexAttribute` is `TANGENTS`, which is documented as "tangent, bitangent and normal, encoded as a quaternion (float4)".  But how exactly do you do that conversion?  I understand that one can put 3 orthonormal vectors into a rotation matrix and convert that to a quaternion, but there are choices of the ordering of the vectors and whether you use row or column vectors.

Second, what if I'm starting from a 3D object format that specifies normal vectors but not a tangent space?  I expect that without that, one would be unable to use certain techniques like anisotropic material.  But how would I best proceed?

**Comment by romainguy** [comment, marked as answer]
You can use our `geometry` library which provides utilities (`SurfaceOrientation`) to generate those quaternions either from a full orthonormal basis or from vertex data. And you should definitely use our implementation to ensure edge cases are properly handled. Excerpt from the API docs:

```
     * Clients provide pointers into their own data, which is synchronously consumed during build().
     * At a minimum, clients must supply a vertex count. They can supply data in any of the
     * following combinations:
     *
     *   1. normals only ........................... not recommended, selects arbitrary orientation
     *   2. normals + tangents ..................... sign of W determines bitangent orientation
     *   3. normals + uvs + positions + indices .... selects Lengyel’s Method
     *   4. positions + indices .................... generates normals for flat shading only
     *
     * Additionally, the client-side data has the following type constraints:
     *
     *  - Normals must be float3
     *  - Tangents must be float4
     *  - UVs must be float2
     *  - Positions must be float3
     *  - Triangles must be uint3 or ushort3
     *
     * Currently, mikktspace is not supported because it requires re-indexing the mesh. Instead
     * we use the method described by Eric Lengyel in "Foundations of Game Engine Development"
     * (Volume 2, Chapter 7).
```

It is of course better to do it offline if possible (like what our `filamesh` tool does).
